plugins/catalog/src/components/CatalogFilter/EntityFilterGroupBar.tsx

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Card,
+  List,
+  ListItemIcon,
+  ListItemText,
+  makeStyles,
+  MenuItem,
+  Theme,
+  Typography,
+} from '@material-ui/core';
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { useEntityListState } from '../../state';
+import { ButtonGroup, useFilterGroups } from './useFilterGroups';
+
+const useStyles = makeStyles<Theme>(theme => ({
+  root: {
+    backgroundColor: 'rgba(0, 0, 0, .11)',
+    boxShadow: 'none',
+  },
+  title: {
+    margin: theme.spacing(1, 0, 0, 1),
+    textTransform: 'uppercase',
+    fontSize: 12,
+    fontWeight: 'bold',
+  },
+  listIcon: {
+    minWidth: 30,
+    color: theme.palette.text.primary,
+  },
+  menuItem: {
+    minHeight: theme.spacing(6),
+  },
+  groupWrapper: {
+    margin: theme.spacing(1, 1, 2, 1),
+  },
+  menuTitle: {
+    fontWeight: 500,
+  },
+}));
+
+type OnChangeCallback = (item: { id: string; label: string }) => void;
+
+type Props = {
+  buttonGroups: ButtonGroup[];
+  initiallySelected: string;
+  onChange?: OnChangeCallback;
+};
+
+/**
+ * The main filter group in the sidebar, toggling owned/starred/all.
+ */
+export const EntityFilterGroupBar = ({
+  buttonGroups,
+  onChange,
+  initiallySelected,
+}: Props) => {
+  const classes = useStyles();
+  const { setFilter } = useEntityListState();
+  const [currentFilter, setCurrentFilter] = useState(initiallySelected);
+
+  const onChangeRef = useRef<OnChangeCallback>();
+
+  useEffect(() => {
+    onChangeRef.current = onChange;
```

Comment:
This is a common-ish pattern for callbacks and initialValue type things. For callbacks, the parent usually does 

```ts
<Foo callback={value => { setState(value) }} />
```

which technically means that a new function instance is generated every single render and passed down. If Foo now has

```ts
useEffect(() => {
  callback(v)
}, [callback, v]);
```

to call the callback when v changes - note that it is forced by the linter to have the callback itself as a dependency parameter. Which will lead to an infinite loop of calling the callback, where the parent sets its state causing a rerender, which... etc.

The realization is that we really have no interest at all in what the actual value is of the callback, we just want to hold on to a box that can contain a reference to the last passed value. Which is exactly what useRef is good at. Now we can depend on the unchanging box, rather than its contents.

Note that we couldn't remove the callback from the dependencies list and silence the linter either. That may or may not work, because we risk calling an old version of the callback instead.

useCallback, in contrast, explicitly generates a new function for you based on inputs changing, which is what we were trying to get away from.

Reply:
@freben Makes sense, thanks for the explanation!

Reply:
Thanks for the excellent explanation @freben! I'll have to admit I don't fully grasp the advanced usages of hooks quite yet, but I get the general concept.